Sources/XCTest/Public/XCTestCase.swift

```diff
@@ -46,8 +46,41 @@ open class XCTestCase: XCTest {
         return 1
     }
 
+    internal var currentWaiter: XCTWaiter?
+
     /// The set of expectations made upon this test case.
-    internal var _allExpectations = [XCTestExpectation]()
+    private var _allExpectations = [XCTestExpectation]()
+
+    internal var expectations: [XCTestExpectation] {
+        return XCTWaiter.subsystemQueue.sync {
+            return _allExpectations
+        }
+    }
+
+    internal func addExpectation(_ expectation: XCTestExpectation) {
+        precondition(Thread.isMainThread, "\(#function) must be called on the main thread")
+        precondition(currentWaiter == nil, "API violation - creating an expectation while already in waiting mode.")
```

Comment:
Yeah, thanks, I realize that will be a behavioral difference. And note that `XCTestExpectation.assertForOverFulfill` is similar in that it causes a `fatalError`, although I opted to always have it be disabled by default unlike Apple XCTest where it is enabled by default for XCTest-generated instances and disabled otherwise.

There just is not a way I am aware of to fully mimic the behavior of ObjC's exceptions here.

Reply:
Ultimately, I suspect we might want to develop the harness to a point that it spawns worker subprocesses to run the actual test code, thus providing a way for test execution to continue and results to not get lost after a failure like this. (Xcode plays a similar role today when running XCTest bundles in that it is able to restart the runner after a crash.)